src/elements/play-assets/play-assets.ts

```diff
@@ -0,0 +1,429 @@
+import {type PropertyValues, ReactiveElement} from 'lit'
+import {customElement, property} from 'lit/decorators.js'
+import {WebAccess, WebStorage} from '@zenfs/dom'
+import {Bubble} from '../../utils/bubble.js'
+import {Zip} from '@zenfs/zip'
+import {
+  hasFileAccessAPI,
+  tryGetFile,
+  tryQueryPermission
+} from '../../utils/file-access-api.js'
+import * as ZenFS from '@zenfs/core'
+import {FileSystem, fs, InMemory} from '@zenfs/core'
+import type {AssetMap} from '@devvit/shared-types/Assets.js'
+import * as IDB from 'idb-keyval'
+
+declare global {
+  interface HTMLElementEventMap {}
+  interface HTMLElementTagNameMap {
+    'play-assets': PlayAssets
+  }
+}
+
+export type AssetFilesystemType = 'virtual' | 'local'
+
+const CACHE_LAST_DIR = 'lastMountedDirectory'
+const CACHE_LAST_ZIP = 'lastMountedArchive'
+
+const MIME: {[ext: string]: string} = {
+  html: 'text/html',
+  htm: 'text/html',
+  jpg: 'image/jpeg',
+  jpeg: 'image/jpeg',
+  png: 'image/png',
+  gif: 'image/gif'
+}
+const DEFAULT_MIME = 'application/octet-stream'
+
+declare global {
+  interface HTMLElementTagNameMap {
+    'play-assets': PlayAssets
+  }
+  interface HTMLElementEventMap {
+    'assets-updated': CustomEvent<void>
+  }
+}
+
+@customElement('play-assets')
+export class PlayAssets extends ReactiveElement {
+  @property({attribute: 'allow-storage', type: Boolean})
+  allowStorage: boolean = false
+
+  @property({attribute: 'filesystem-type', reflect: true, type: String})
+  filesystemType: AssetFilesystemType = 'virtual'
+
+  @property({attribute: false})
+  directoryHandle: FileSystemDirectoryHandle | undefined
+
+  @property({attribute: false})
+  directoryName: string | undefined
+
+  @property({attribute: false})
+  archiveHandle: FileSystemFileHandle | File | undefined
+
+  @property({attribute: false})
+  archiveFilename: string | undefined
+
+  @property({attribute: false})
```

Comment:
why wouldn't we want an attribute for a non-object? I've been operating on the assumption that the DOM is declarative and it's easy to debug attributes.

Reply:
I've refactored again away from this, but for the sake of discussion :smile: 

Rationale:
`@state` is for private/protected reactive state, `@property` is for publicly accessible reactive state.  These are dynamic properties that are updated as you mess with the mount settings and I want UI to update as they're modified without explicitly setting up event listeners and local state to consumers, but don't want these to be visible attributes.